### DISCUSSION TITLE
Debt cap and liquidation in epoch4

### DISCUSSION THREAD

**Discussion by kevingzhang** [tearust/teaproject#72] [Ideas]
One of the design issue of epoch3 is the Coffee debt.

For many contestants, they feel it is so easy to borrow coffee and invest into the tea mining as if there is no cost. Although we mentioned so many times that you have to pay off debt to get qualified to be the winner. There are about half of the contestants cannot pay off at the end.

Another problem caused by that is the TEA/Coffee exchange rate at the end. Since there are many contestants trying to pay off their debt at the end, they sell their tapp token and TEA for Coffee. That TEA/Coffee exchange rate dropped from around 2 to 0.6. Of course, this is not what we expected in real life. 

It is very important that we test this situation in the testing environment so that we can improve the design to prevent it from happening in real life.

In epoch4, I propose the following changes and following new concepts.

# Debt cap
Debt_cap = max(0.5 * your_net_assets, time_based_debt_cap).
time_based_debt_cap = position_or_zero(200,000 - current_block_height) / 10 

So your debt cap is determined by two variables: 
- Your net assets which is (your gross assets - debt)
- Time.

At the beginning, you will have the initial 20k COFFEE. As time goes by, the time_based_debt_cap drops but your net assets growth (I hope so). The max value of those two is your current debt cap.

# Liquidation
What if my debt is over my current debt cap?

A mandatory liquidation takes place when your debt is greater than your debt cap. In the UI, we will show those two numbers. If you are in liquidation mode, those two numbers will be shown in red. Every payment to your TEA account will be sold to COFFEE and pay off your exceed debt. 

Every payment includes all your income from mining, hosting, FaaS, sell of token etc.

When your debt is lower than the debt cap, you are out of the liquidation mode.

# When to check the debt and put into liquidation mode?

Every 1000 blocks, system will check your debt and determine if you are in liquidation mode or not. You can fool the system by reduce the debt every 1000 blocks, that is allowed. But if you cannot pay off your debt at that moment, you will be in the liquidation mode.

# Borrow limit
The debt cap is always checked everything you try to borrow COFFEE.

# When the block number exceed 200,000
That would be the last 2 days of the epoch. Your debt cap is determined solely by your assets. If you have enough asset, you can still borrow COFFEE. 

# Do I need to pay off all COFFEE debt at the end of the epoch to be qualified to be winner?

NO, you do not need to. As long as your debt amount lower than your current debt cap, you are qualified. So you do not need to rush to sell your assets for COFFEE.

**Comment by geopor** [comment]
**Debt Cap** - another way to state this complicated formula:
Contestants can borrow up to 20,000 COFFEE at the very beginning. At the contest continues, this debt ceiling goes lower. 

- At block height 10,000, which would be 16 hours and 40 minutes, it drops to 19,000.
- At block height 20,000, which would be 1 day 9 hours and 20 minutes, it drops to 18,000.
- At block height 30,000, which would be 2 days and 2 hours, the debt cap drops to 17,000.

If a contestant has borrowed let's say 16,500 COFFEE and after a couple days of .02% interest (30,000 blocks) their debt is now 17,173 COFFEE. But at block height 30,000 they need to be below 17,000 COFFEE debt.
To avoid liquidation of their assets (i.e. forcefully sold) to get them under 17,000 COFFEE debt, they can see if 50% of their (assets - debt) is above this number.

## Some discussion points

- The more you borrow, the sooner the debt clock catches up to you. If you borrow 12,500 COFFEE, you know that after 4 days the interest puts your debt slightly above 14,000, and 4 days is when the debt ceiling lowers to 14,000. So it's probably liquidation time unless your assets are worth more than 44,000 COFFEE: 50% of (44K assets - 14K debt) = 15,000.
- The 50% (assets - debt) is not likely to be more than the debt until the debt cap has dropped much further into the contest. It likely won't save someone in the earlier stages.
- At the end of epoch 3, contestants sold their TApp tokens. This benefits them and, by dropping the value of the TApp tokens on the bonding curve, it hurts other token holders. This could be a double whammy for people who choose to hold TApp tokens at the end: more cunning contestants will sell their TApp tokens to secure their value, thereby dropping the price. This drop in price affects other holders but could also crater their (asset - debt) value, sparking liquidations. This would be bad economic design if this happened as the system would be destroying economic value for no good reason. There may be a need to stop trading during the final days of the contest to avoid this scenario, though it would have to be a fuzzy deadline otherwise contestants will sell right up to the deadline.
- Because of **theta** there will always be a time lag before the asset value of TApp tokens (which are valued at the sell price) catch up to the amount of debt used to purchase them (which is measured at the buy price). The debt grows by .02% interest per 100 blocks, while the TApp tokens grow by appreciation when other people buy them. If theta is too high (even .3 might be too high), the spread will make it such that the token appreciation (valued at the sell price) has a longer time to catch up to where they bought it at. Ideally theta values are kept low, at least below .3 but creators can choose their own values.

**Reply by kevingzhang** [reply]
> **Debt Cap** - another way to state this complicated formula:
> Contestants can borrow up to 20,000 COFFEE at the very beginning. At the contest continues, this debt ceiling goes lower.
> 
> * At block height 10,000, which would be 16 hours and 40 minutes, it drops to 19,000.
> * At block height 20,000, which would be 1 day 9 hours and 20 minutes, it drops to 18,000.
> * At block height 30,000, which would be 2 days and 2 hours, the debt cap drops to 17,000.
> 

For humans, this is easier, but for computers, the formula is easier. What we can do is to use the formula in a computer, but showing some kind of human-readable rules like this. As long as the debt cap is higher than what humans can read, they will accept.

> If a contestant has borrowed let's say 16,500 COFFEE and after a couple days of .02% interest (30,000 blocks) their debt is now 17,173 COFFEE. But at block height 30,000 they need to be below 17,000 COFFEE debt.
> To avoid liquidation of their assets (i.e. forcefully sold) to get them under 17,000 COFFEE debt, they can see if 50% of their (assets - debt) is above this number.

We will have a constant number showing in the first page of wallet. So the user will see the up-to-date debt cap every minute. They will know how far they are from the debt cap. 

Also, after the discussion with devs, it would be pretty hard to steal money away from the user account. But we can add a new penalty. I do not know what it should be called. "Overdraft penalty?" or something else? 
Every 1000 blocks, the system will check is there anyone has to exceed the debt cap. If so, add one violation in the blockchain. On the UI, especially the leaderboard, showing one violation. The next 1000 blocks, it will check again. if still beyond the limit, then the second violation. The violation will turn to a T penalty in the leader board. For example, one violation is 10T penalty. This is an easy implementation.
 

> 
> ## Some discussion points
> * The more you borrow, the sooner the debt clock catches up to you. If you borrow 12,500 COFFEE, you know that after 4 days the interest puts your debt slightly above 14,000, and 4 days is when the debt ceiling lowers to 14,000. So it's probably liquidation time unless your assets are worth more than 44,000 COFFEE: 50% of (44K assets - 14K debt) = 15,000.
> * The 50% (assets - debt) is not likely to be more than the debt until the debt cap has dropped much further into the contest. It likely won't save someone in the earlier stages.
> * At the end of epoch 3, contestants sold their TApp tokens. This benefits them and, by dropping the value of the TApp tokens on the bonding curve, it hurts other token holders. This could be a double whammy for people who choose to hold TApp tokens at the end: more cunning contestants will sell their TApp tokens to secure their value, thereby dropping the price. This drop in price affects other holders but could also crater their (asset - debt) value, sparking liquidations. This would be bad economic design if this happened as the system would be destroying economic value for no good reason. There may be a need to stop trading during the final days of the contest to avoid this scenario, though it would have to be a fuzzy deadline otherwise contestants will sell right up to the deadline.

If we do not force them to sell for coffee to pay off debt, they have no incentive to sell the token. The token has value and also included in the asset. If they sell, everyone including themselves will lose value. So the new design actually fixed this problem. They do not need to sell, and sell make everyone's valuation lower.

> * Because of **theta** there will always be a time lag before the asset value of TApp tokens (which are valued at the sell price) catch up to the amount of debt used to purchase them (which is measured at the buy price). The debt grows by .02% interest per 100 blocks, while the TApp tokens grow by appreciation when other people buy them. If theta is too high (even .3 might be too high), the spread will make it such that the token appreciation (valued at the sell price) has a longer time to catch up to where they bought it at. Ideally theta values are kept low, at least below .3 but creators can choose their own values.

**Reply by geopor** [reply]
Overdraft penalty sounds good. I like that design better than liquidating assets, as liquidations are very hard on participants and will sink their debt ratios even more.

**Reply by geopor** [reply]
Also I hope this can be implemented during the accelerated testing phase of epoch 4 so that testers can stress test different strategies and what are the likely debt outcomes. It would be good to get some data so we can tweak before releasing as epoch contest is 2 weeks longs.

**Reply by kevingzhang** [reply]
I hope too, but our development is left behind the schedule. I do not know how long this feature can be completed since this is added the last minute. let's see how fast we can get

**Reply by kevingzhang** [reply]
One more possible feature we can do is to kill an TApp. Only owner can do this. Once owner decided to kill this tapp, all existing investor will share the TEA token in the reservation pool based on their holding tokens.
Regarding the owner own theta part, my current idea is to let owner to return those theta back to all the token holders.

If the owner can pay back, then the investors lose nothing.

So here are questions:
- Why would the tapp owner do this?
- What if the tapp owner is willing to kill the tapp but not rich enough to pay off all theta? do we still allow the owner to kill the tapp? If he cannot kill the tapp, he can just sell all his own token and walk way, is there anything we can do with this?

**Reply by geopor** [reply]
In real life, devs abandon projects all the time as often happens when they get busy with other opportunities. 

I don't understand why they have to pay back the theta portion of the bonding curve. This portion is also referred to as the **Funding Pool** and is used as capital for the project can pay its developers. Forcing pay back of the theta portion is like asking them to pay back their salaries.

**Reply by kevingzhang** [reply]
That is true. there is no way to pay back the theta portion for sure. So the investors will just take all of the risk, but that is normal. Investment has risk. If the developer just mean to take away this portion of money, they can do at any time. There is no way to prevent them doing so. 

I am just thinking the case in the epoch4. Someone post a video, create a Teafluencer. Other invest to it. he sell all the initial token and walk away. If this happen, the remaining investor can either sell and leave , or still consider the video will make new views and new tips. So the key point is not whether the dev walk away or not. The key point is whether the video can make future tips or not. 

If the dev (owner) think this video can no longer make new tips, anyone, including himself will walk away. So it is no need to explicitly close/kill this app. It will die anyway.

My only reason to have a kill feature is that what if the dev create the Teaflucner with wrong theta, and he wants to correct it. So far, since there are investor's money in the pool, we need a way to refund the investor and restart a new Teaflucncer app again. 

It looks there is no way to fix the problem of changing theta or kill the app. If we do not have any new solution, we will have to tell the devs that there is no way to change once the app is created. If the theta value is not good, the only way is to let the tapp die.

**Reply by geopor** [reply]
Ok I see. Yes most devs who "abandon" a project will just sell their stake and leave the app running. Like you said, the remaining investors might be financially hurt by this action but long term they could still end up profitable if the underlying video does well. 

In terms of if the dev made a mistake, the ability to kill the TApp & refund theta would be useful. Hopefully most errors would be caught early and so there would be minimal theta that needed to be refunded to investors.
